File: engines/engine_comboBuilder.py
from __future__ import annotations

from typing import TYPE_CHECKING

from engines.engine_log import DEBUG, INFO
from utilities.constant import ComboType, Direction
from utilities.object import ContractData, Leg

if TYPE_CHECKING:
    from engines.engine_strategy import OptionStrategyEngine
    from utilities.portfolio import OptionData


class ComboBuilderEngine:
    """Engine responsible for building combo orders and managing combo strategies."""

    def __init__(self, option_strategy_engine: OptionStrategyEngine) -> None:
        self.option_strategy_engine = option_strategy_engine

    def write_log(self, msg: str, level: int = INFO) -> None:
        """Write log message through the strategy engine."""
        self.option_strategy_engine.write_log(msg, level)

    # ====================== LEG CREATION ======================

    def create_leg(self, option: OptionData, direction: Direction, volume: int, price: float | None = None) -> Leg:
        """Create a leg for combo orders from OptionData by resolving its contract."""
        contract: ContractData | None = self.option_strategy_engine.get_contract(option.symbol)
        if not contract:
            raise ValueError(f"Contract not found for option: {option.symbol}")

        leg = Leg(
            con_id=contract.con_id or 0,
            symbol=contract.symbol,
            exchange=contract.exchange,
            direction=direction,
            ratio=volume,
            price=price,
            gateway_name="IB",
            trading_class=contract.trading_class,
        )

        return leg

    def combo_builder(
        self, option_data: dict[str, OptionData], combo_type: ComboType, direction: Direction, volume: int
    ) -> tuple[list[Leg], str]:
        """Create combo from option data based on combo type.

        Args:
            option_data: Dictionary mapping leg names to OptionData objects
                - For STRADDLE: {"call": OptionData, "put": OptionData}
                - For STRANGLE: {"call": OptionData, "put": OptionData}
                - For SPREAD: {"long_leg": OptionData, "short_leg": OptionData}
                - For DIAGONAL_SPREAD: {"long_leg": OptionData, "short_leg": OptionData}
                - For RATIO_SPREAD: {"long_leg": OptionData, "short_leg": OptionData, "ratio": int}
                - For BUTTERFLY: {"body": OptionData, "wing1": OptionData, "wing2": OptionData}
                - For INVERSE_BUTTERFLY: {"body": OptionData, "wing1": OptionData, "wing2": OptionData}
                - For IRON_BUTTERFLY: {"put_wing": OptionData, "body": OptionData, "call_wing": OptionData}
                - For IRON_CONDOR: {"put_lower": OptionData, "put_upper": OptionData, "call_lower": OptionData, "call_upper": OptionData}
                - For CONDOR: {"long_put": OptionData, "short_put": OptionData, "short_call": OptionData, "long_call": OptionData}
                - For BOX_SPREAD: {"long_call": OptionData, "short_call": OptionData, "short_put": OptionData, "long_put": OptionData}
                - For RISK_REVERSAL: {"long_leg": OptionData, "short_leg": OptionData}
                - For CUSTOM: {"leg1": OptionData, "leg2": OptionData, ...}
        """

        if combo_type == ComboType.STRADDLE:
            return self.straddle(option_data, direction, volume)

        elif combo_type == ComboType.STRANGLE:
            return self.strangle(option_data, direction, volume)

        elif combo_type == ComboType.IRON_CONDOR:
            return self.iron_condor(option_data, direction, volume)

        elif combo_type == ComboType.RISK_REVERSAL:
            return self.risk_reversal(option_data, direction, volume)

        elif combo_type == ComboType.SPREAD:
            return self.spread(option_data, direction, volume)

        elif combo_type == ComboType.DIAGONAL_SPREAD:
            return self.diagonal_spread(option_data, direction, volume)

        elif combo_type == ComboType.RATIO_SPREAD:
            return self.ratio_spread(option_data, direction, volume)

        elif combo_type == ComboType.BUTTERFLY:
            return self.butterfly(option_data, direction, volume)

        elif combo_type == ComboType.INVERSE_BUTTERFLY:
            return self.inverse_butterfly(option_data, direction, volume)

        elif combo_type == ComboType.IRON_BUTTERFLY:
            return self.iron_butterfly(option_data, direction, volume)

        elif combo_type == ComboType.CONDOR:
            return self.condor(option_data, direction, volume)

        elif combo_type == ComboType.BOX_SPREAD:
            return self.box_spread(option_data, direction, volume)

        elif combo_type == ComboType.CUSTOM:
            return self.custom(option_data, direction, volume)

        else:
            raise ValueError(f"Unsupported combo type: {combo_type}")

    # ====================== COMBO BUILDERS ======================

    def straddle(self, option_data: dict[str, OptionData], direction: Direction, volume: int) -> tuple[list[Leg], str]:
        """Build straddle combo: ATM call + ATM put."""

        legs = [self.create_leg(option_data["call"], direction, volume), self.create_leg(option_data["put"], direction, volume)]
        sig = self.generate_combo_signature(legs)
        return legs, sig

    def strangle(self, option_data: dict[str, OptionData], direction: Direction, volume: int) -> tuple[list[Leg], str]:
        """Build strangle combo: OTM call + OTM put."""

        legs = [self.create_leg(option_data["call"], direction, volume), self.create_leg(option_data["put"], direction, volume)]
        sig = self.generate_combo_signature(legs)
        return legs, sig

    def iron_condor(self, option_data: dict[str, OptionData], direction: Direction, volume: int) -> tuple[list[Leg], str]:
        """Build iron condor combo: Long put + Short put + Short call + Long call."""
        sign = 1 if direction == Direction.SHORT else -1

        legs = [
            self.create_leg(option_data["put_lower"], Direction.LONG if sign > 0 else Direction.SHORT, volume),
            self.create_leg(option_data["put_upper"], Direction.SHORT if sign > 0 else Direction.LONG, volume),
            self.create_leg(option_data["call_lower"], Direction.SHORT if sign > 0 else Direction.LONG, volume),
            self.create_leg(option_data["call_upper"], Direction.LONG if sign > 0 else Direction.SHORT, volume),
        ]

        sig = self.generate_combo_signature(legs)
        return legs, sig

    def risk_reversal(self, option_data: dict[str, OptionData], direction: Direction, volume: int) -> tuple[list[Leg], str]:
        """Build risk reversal combo: Long call + Short put."""

        sign = 1 if direction == Direction.SHORT else -1

        legs = [
            self.create_leg(option_data["long_leg"], Direction.LONG if sign > 0 else Direction.SHORT, volume),
            self.create_leg(option_data["short_leg"], Direction.SHORT if sign > 0 else Direction.LONG, volume),
        ]
        sig = self.generate_combo_signature(legs)

        return legs, sig

    def custom(self, option_data: dict[str, OptionData], direction: Direction, volume: int) -> tuple[list[Leg], str]:
        """Build custom combo from option data Used to close positions"""
        legs: list[Leg] = []
        for option in option_data.values():
            # For closing positions, use the specified direction for all legs
            # This is correct - when closing a LONG position, we SHORT all legs
            leg = self.create_leg(option, direction, volume)
            legs.append(leg)
            # Debug: Log leg creation for closing positions
            self.write_log(f"Custom Combo Leg: {leg.symbol} | Direction: {leg.direction} | Volume: {leg.ratio}", DEBUG)
        sig = self.generate_combo_signature(legs)
        return legs, sig

    def spread(self, option_data: dict[str, OptionData], direction: Direction, volume: int) -> tuple[list[Leg], str]:
        """Build vertical/calendar spread combo: Long leg + Short leg."""
        sign = 1 if direction == Direction.LONG else -1

        legs = [
            self.create_leg(option_data["long_leg"], Direction.LONG if sign > 0 else Direction.SHORT, volume),
            self.create_leg(option_data["short_leg"], Direction.SHORT if sign > 0 else Direction.LONG, volume),
        ]
        sig = self.generate_combo_signature(legs)
        return legs, sig

    def diagonal_spread(self, option_data: dict[str, OptionData], direction: Direction, volume: int) -> tuple[list[Leg], str]:
        """Build diagonal spread combo: Long leg + Short leg (different strikes and expirations)."""
        sign = 1 if direction == Direction.LONG else -1

        legs = [
            self.create_leg(option_data["long_leg"], Direction.LONG if sign > 0 else Direction.SHORT, volume),
            self.create_leg(option_data["short_leg"], Direction.SHORT if sign > 0 else Direction.LONG, volume),
        ]
        sig = self.generate_combo_signature(legs)
        return legs, sig

    def ratio_spread(self, option_data: dict[str, OptionData], direction: Direction, volume: int) -> tuple[list[Leg], str]:
        """Build ratio spread combo: Long leg + Multiple short legs."""
        sign = 1 if direction == Direction.LONG else -1
        ratio = option_data.get("ratio", 2)  # Default 1:2 ratio

        legs = [
            self.create_leg(option_data["long_leg"], Direction.LONG if sign > 0 else Direction.SHORT, volume),
            self.create_leg(option_data["short_leg"], Direction.SHORT if sign > 0 else Direction.LONG, volume * ratio),
        ]
        sig = self.generate_combo_signature(legs)
        return legs, sig

    def butterfly(self, option_data: dict[str, OptionData], direction: Direction, volume: int) -> tuple[list[Leg], str]:
        """Build butterfly combo: Long body + Short wings."""
        sign = 1 if direction == Direction.LONG else -1

        legs = [
            self.create_leg(option_data["body"], Direction.LONG if sign > 0 else Direction.SHORT, volume),
            self.create_leg(option_data["wing1"], Direction.SHORT if sign > 0 else Direction.LONG, volume),
            self.create_leg(option_data["wing2"], Direction.SHORT if sign > 0 else Direction.LONG, volume),
        ]
        sig = self.generate_combo_signature(legs)
        return legs, sig

    def inverse_butterfly(self, option_data: dict[str, OptionData], direction: Direction, volume: int) -> tuple[list[Leg], str]:
        """Build inverse butterfly combo: Short body + Long wings."""
        sign = 1 if direction == Direction.LONG else -1

        legs = [
            self.create_leg(option_data["body"], Direction.SHORT if sign > 0 else Direction.LONG, volume),
            self.create_leg(option_data["wing1"], Direction.LONG if sign > 0 else Direction.SHORT, volume),
            self.create_leg(option_data["wing2"], Direction.LONG if sign > 0 else Direction.SHORT, volume),
        ]
        sig = self.generate_combo_signature(legs)
        return legs, sig

    def iron_butterfly(self, option_data: dict[str, OptionData], direction: Direction, volume: int) -> tuple[list[Leg], str]:
        """Build iron butterfly combo: Long put wing + Short body + Long call wing."""
        sign = 1 if direction == Direction.LONG else -1

        legs = [
            self.create_leg(option_data["put_wing"], Direction.LONG if sign > 0 else Direction.SHORT, volume),
            self.create_leg(option_data["body"], Direction.SHORT if sign > 0 else Direction.LONG, volume),
            self.create_leg(option_data["call_wing"], Direction.LONG if sign > 0 else Direction.SHORT, volume),
        ]
        sig = self.generate_combo_signature(legs)
        return legs, sig

    def condor(self, option_data: dict[str, OptionData], direction: Direction, volume: int) -> tuple[list[Leg], str]:
        """Build condor combo: Long put + Short put + Short call + Long call."""
        sign = 1 if direction == Direction.LONG else -1

        legs = [
            self.create_leg(option_data["long_put"], Direction.LONG if sign > 0 else Direction.SHORT, volume),
            self.create_leg(option_data["short_put"], Direction.SHORT if sign > 0 else Direction.LONG, volume),
            self.create_leg(option_data["short_call"], Direction.SHORT if sign > 0 else Direction.LONG, volume),
            self.create_leg(option_data["long_call"], Direction.LONG if sign > 0 else Direction.SHORT, volume),
        ]
        sig = self.generate_combo_signature(legs)
        return legs, sig

    def box_spread(self, option_data: dict[str, OptionData], direction: Direction, volume: int) -> tuple[list[Leg], str]:
        """Build box spread combo: Long call + Short call + Short put + Long put."""
        sign = 1 if direction == Direction.LONG else -1

        legs = [
            self.create_leg(option_data["long_call"], Direction.LONG if sign > 0 else Direction.SHORT, volume),
            self.create_leg(option_data["short_call"], Direction.SHORT if sign > 0 else Direction.LONG, volume),
            self.create_leg(option_data["short_put"], Direction.SHORT if sign > 0 else Direction.LONG, volume),
            self.create_leg(option_data["long_put"], Direction.LONG if sign > 0 else Direction.SHORT, volume),
        ]
        sig = self.generate_combo_signature(legs)
        return legs, sig

    @staticmethod
    def generate_combo_signature(legs: list[Leg]) -> str:
        """Build unique signature string from multiple legs."""
        leg_signatures = []
        for leg in legs:
            if not leg.symbol:
                continue
            parts = leg.symbol.split("-")
            expiry, opt_type, strike = parts[1:4]
            leg_signatures.append(f"{expiry}{opt_type}{strike}")
        return "-".join(sorted(leg_signatures))

File: Otrader_cpp/core/engine_combo_builder.cpp
/**
 * Shared ComboBuilder implementation; logs output as LogIntent (out_logs).
 */

#include "engine_combo_builder.hpp"
#include <algorithm>
#include <sstream>

namespace engines {

auto ComboBuilderEngine::create_leg(const utilities::OptionData& option,
    utilities::Direction direction, int volume, std::optional<double> price,
    const ComboGetContractFn& get_contract_fn) -> utilities::Leg {
    const utilities::ContractData* contract = nullptr;
    if (get_contract_fn) {
        contract = get_contract_fn(option.symbol);
    } else if (current_get_contract_) {
        contract = current_get_contract_(option.symbol);
}
    if (contract == nullptr) {
        throw std::runtime_error("Contract not found for option: " + option.symbol);
    }
    utilities::Leg leg;
    leg.con_id = contract->con_id.value_or(0);
    leg.symbol = contract->symbol;
    leg.exchange = contract->exchange;
    leg.direction = direction;
    leg.ratio = volume;
    leg.price = price;
    leg.gateway_name = "IB";
    leg.trading_class = contract->trading_class;
    return leg;
}

auto ComboBuilderEngine::combo_builder(
    const std::unordered_map<std::string, utilities::OptionData*>& option_data,
    utilities::ComboType combo_type, utilities::Direction direction, int volume,
    ComboGetContractFn get_contract_fn, std::vector<utilities::LogData>* out_logs) -> std::pair<std::vector<utilities::Leg>, std::string> {
    current_get_contract_ = std::move(get_contract_fn);
    current_out_logs_ = out_logs;
    auto result = combo_builder_impl(option_data, combo_type, direction, volume);
    current_get_contract_ = nullptr;
    current_out_logs_ = nullptr;
    return result;
}

auto ComboBuilderEngine::combo_builder_impl(
    const std::unordered_map<std::string, utilities::OptionData*>& option_data,
    utilities::ComboType combo_type, utilities::Direction direction, int volume) -> std::pair<std::vector<utilities::Leg>, std::string> {
    switch (combo_type) {
        case utilities::ComboType::STRADDLE: return straddle(option_data, direction, volume);
        case utilities::ComboType::STRANGLE: return strangle(option_data, direction, volume);
        case utilities::ComboType::IRON_CONDOR: return iron_condor(option_data, direction, volume);
        case utilities::ComboType::RISK_REVERSAL: return risk_reversal(option_data, direction, volume);
        case utilities::ComboType::SPREAD: return spread(option_data, direction, volume);
        case utilities::ComboType::DIAGONAL_SPREAD: return diagonal_spread(option_data, direction, volume);
        case utilities::ComboType::RATIO_SPREAD: return ratio_spread(option_data, direction, volume);
        case utilities::ComboType::BUTTERFLY: return butterfly(option_data, direction, volume);
        case utilities::ComboType::INVERSE_BUTTERFLY: return inverse_butterfly(option_data, direction, volume);
        case utilities::ComboType::IRON_BUTTERFLY: return iron_butterfly(option_data, direction, volume);
        case utilities::ComboType::CONDOR: return condor(option_data, direction, volume);
        case utilities::ComboType::BOX_SPREAD: return box_spread(option_data, direction, volume);
        case utilities::ComboType::CUSTOM: return custom(option_data, direction, volume);
        default: throw std::runtime_error("Unsupported combo type");
    }
}

auto ComboBuilderEngine::straddle(
    const std::unordered_map<std::string, utilities::OptionData*>& option_data,
    utilities::Direction direction, int volume) -> std::pair<std::vector<utilities::Leg>, std::string> {
    auto it_c = option_data.find("call");
    auto it_p = option_data.find("put");
    if (it_c == option_data.end() || it_p == option_data.end()) {
        throw std::runtime_error("straddle requires 'call' and 'put'");
}
    std::vector<utilities::Leg> legs = {
        create_leg(*it_c->second, direction, volume),
        create_leg(*it_p->second, direction, volume)
    };
    return {legs, generate_combo_signature(legs)};
}

auto ComboBuilderEngine::strangle(
    const std::unordered_map<std::string, utilities::OptionData*>& option_data,
    utilities::Direction direction, int volume) -> std::pair<std::vector<utilities::Leg>, std::string> {
    auto it_c = option_data.find("call");
    auto it_p = option_data.find("put");
    if (it_c == option_data.end() || it_p == option_data.end()) {
        throw std::runtime_error("strangle requires 'call' and 'put'");
}
    std::vector<utilities::Leg> legs = {
        create_leg(*it_c->second, direction, volume),
        create_leg(*it_p->second, direction, volume)
    };
    return {legs, generate_combo_signature(legs)};
}

auto ComboBuilderEngine::iron_condor(
    const std::unordered_map<std::string, utilities::OptionData*>& option_data,
    utilities::Direction direction, int volume) -> std::pair<std::vector<utilities::Leg>, std::string> {
    int sign = (direction == utilities::Direction::SHORT) ? 1 : -1;
    auto *pl = option_data.at("put_lower");
    auto *pu = option_data.at("put_upper");
    auto *cl = option_data.at("call_lower");
    auto *cu = option_data.at("call_upper");
    std::vector<utilities::Leg> legs = {
        create_leg(*pl, sign > 0 ? utilities::Direction::LONG : utilities::Direction::SHORT, volume),
        create_leg(*pu, sign > 0 ? utilities::Direction::SHORT : utilities::Direction::LONG, volume),
        create_leg(*cl, sign > 0 ? utilities::Direction::SHORT : utilities::Direction::LONG, volume),
        create_leg(*cu, sign > 0 ? utilities::Direction::LONG : utilities::Direction::SHORT, volume),
    };
    return {legs, generate_combo_signature(legs)};
}

auto ComboBuilderEngine::risk_reversal(
    const std::unordered_map<std::string, utilities::OptionData*>& option_data,
    utilities::Direction direction, int volume) -> std::pair<std::vector<utilities::Leg>, std::string> {
    int sign = (direction == utilities::Direction::SHORT) ? 1 : -1;
    auto *ll = option_data.at("long_leg");
    auto *sl = option_data.at("short_leg");
    std::vector<utilities::Leg> legs = {
        create_leg(*ll, sign > 0 ? utilities::Direction::LONG : utilities::Direction::SHORT, volume),
        create_leg(*sl, sign > 0 ? utilities::Direction::SHORT : utilities::Direction::LONG, volume),
    };
    return {legs, generate_combo_signature(legs)};
}

auto ComboBuilderEngine::custom(
    const std::unordered_map<std::string, utilities::OptionData*>& option_data,
    utilities::Direction direction, int volume) -> std::pair<std::vector<utilities::Leg>, std::string> {
    std::vector<utilities::Leg> legs;
    for (const auto& kv : option_data) {
        legs.push_back(create_leg(*kv.second, direction, volume));
        if (current_out_logs_ != nullptr) {
            utilities::LogData log;
            log.msg = "Custom Combo Leg: " + legs.back().symbol.value_or("") +
                " | Direction: " + std::to_string(static_cast<int>(direction)) + " | Volume: " + std::to_string(legs.back().ratio);
            log.level = 10;
            log.gateway_name = "Combo";
            current_out_logs_->push_back(log);
        }
    }
    return {legs, generate_combo_signature(legs)};
}

auto ComboBuilderEngine::spread(
    const std::unordered_map<std::string, utilities::OptionData*>& option_data,
    utilities::Direction direction, int volume) -> std::pair<std::vector<utilities::Leg>, std::string> {
    int sign = (direction == utilities::Direction::LONG) ? 1 : -1;
    auto *ll = option_data.at("long_leg");
    auto *sl = option_data.at("short_leg");
    std::vector<utilities::Leg> legs = {
        create_leg(*ll, sign > 0 ? utilities::Direction::LONG : utilities::Direction::SHORT, volume),
        create_leg(*sl, sign > 0 ? utilities::Direction::SHORT : utilities::Direction::LONG, volume),
    };
    return {legs, generate_combo_signature(legs)};
}

auto ComboBuilderEngine::diagonal_spread(
    const std::unordered_map<std::string, utilities::OptionData*>& option_data,
    utilities::Direction direction, int volume) -> std::pair<std::vector<utilities::Leg>, std::string> {
    int sign = (direction == utilities::Direction::LONG) ? 1 : -1;
    auto *ll = option_data.at("long_leg");
    auto *sl = option_data.at("short_leg");
    std::vector<utilities::Leg> legs = {
        create_leg(*ll, sign > 0 ? utilities::Direction::LONG : utilities::Direction::SHORT, volume),
        create_leg(*sl, sign > 0 ? utilities::Direction::SHORT : utilities::Direction::LONG, volume),
    };
    return {legs, generate_combo_signature(legs)};
}

auto ComboBuilderEngine::ratio_spread(
    const std::unordered_map<std::string, utilities::OptionData*>& option_data,
    utilities::Direction direction, int volume) -> std::pair<std::vector<utilities::Leg>, std::string> {
    int sign = (direction == utilities::Direction::LONG) ? 1 : -1;
    int ratio = 2;  // default 1:2 ratio, same as Python
    auto *ll = option_data.at("long_leg");
    auto *sl = option_data.at("short_leg");
    std::vector<utilities::Leg> legs = {
        create_leg(*ll, sign > 0 ? utilities::Direction::LONG : utilities::Direction::SHORT, volume),
        create_leg(*sl, sign > 0 ? utilities::Direction::SHORT : utilities::Direction::LONG, volume * ratio),
    };
    return {legs, generate_combo_signature(legs)};
}

auto ComboBuilderEngine::butterfly(
    const std::unordered_map<std::string, utilities::OptionData*>& option_data,
    utilities::Direction direction, int volume) -> std::pair<std::vector<utilities::Leg>, std::string> {
    int sign = (direction == utilities::Direction::LONG) ? 1 : -1;
    auto *body = option_data.at("body");
    auto *w1 = option_data.at("wing1");
    auto *w2 = option_data.at("wing2");
    std::vector<utilities::Leg> legs = {
        create_leg(*body, sign > 0 ? utilities::Direction::LONG : utilities::Direction::SHORT, volume),
        create_leg(*w1, sign > 0 ? utilities::Direction::SHORT : utilities::Direction::LONG, volume),
        create_leg(*w2, sign > 0 ? utilities::Direction::SHORT : utilities::Direction::LONG, volume),
    };
    return {legs, generate_combo_signature(legs)};
}

auto ComboBuilderEngine::inverse_butterfly(
    const std::unordered_map<std::string, utilities::OptionData*>& option_data,
    utilities::Direction direction, int volume) -> std::pair<std::vector<utilities::Leg>, std::string> {
    int sign = (direction == utilities::Direction::LONG) ? 1 : -1;
    auto *body = option_data.at("body");
    auto *w1 = option_data.at("wing1");
    auto *w2 = option_data.at("wing2");
    std::vector<utilities::Leg> legs = {
        create_leg(*body, sign > 0 ? utilities::Direction::SHORT : utilities::Direction::LONG, volume),
        create_leg(*w1, sign > 0 ? utilities::Direction::LONG : utilities::Direction::SHORT, volume),
        create_leg(*w2, sign > 0 ? utilities::Direction::LONG : utilities::Direction::SHORT, volume),
    };
    return {legs, generate_combo_signature(legs)};
}

auto ComboBuilderEngine::iron_butterfly(
    const std::unordered_map<std::string, utilities::OptionData*>& option_data,
    utilities::Direction direction, int volume) -> std::pair<std::vector<utilities::Leg>, std::string> {
    int sign = (direction == utilities::Direction::LONG) ? 1 : -1;
    auto *pw = option_data.at("put_wing");
    auto *body = option_data.at("body");
    auto *cw = option_data.at("call_wing");
    std::vector<utilities::Leg> legs = {
        create_leg(*pw, sign > 0 ? utilities::Direction::LONG : utilities::Direction::SHORT, volume),
        create_leg(*body, sign > 0 ? utilities::Direction::SHORT : utilities::Direction::LONG, volume),
        create_leg(*cw, sign > 0 ? utilities::Direction::LONG : utilities::Direction::SHORT, volume),
    };
    return {legs, generate_combo_signature(legs)};
}

auto ComboBuilderEngine::condor(
    const std::unordered_map<std::string, utilities::OptionData*>& option_data,
    utilities::Direction direction, int volume) -> std::pair<std::vector<utilities::Leg>, std::string> {
    int sign = (direction == utilities::Direction::LONG) ? 1 : -1;
    auto *lp = option_data.at("long_put");
    auto *sp = option_data.at("short_put");
    auto *sc = option_data.at("short_call");
    auto *lc = option_data.at("long_call");
    std::vector<utilities::Leg> legs = {
        create_leg(*lp, sign > 0 ? utilities::Direction::LONG : utilities::Direction::SHORT, volume),
        create_leg(*sp, sign > 0 ? utilities::Direction::SHORT : utilities::Direction::LONG, volume),
        create_leg(*sc, sign > 0 ? utilities::Direction::SHORT : utilities::Direction::LONG, volume),
        create_leg(*lc, sign > 0 ? utilities::Direction::LONG : utilities::Direction::SHORT, volume),
    };
    return {legs, generate_combo_signature(legs)};
}

auto ComboBuilderEngine::box_spread(
    const std::unordered_map<std::string, utilities::OptionData*>& option_data,
    utilities::Direction direction, int volume) -> std::pair<std::vector<utilities::Leg>, std::string> {
    int sign = (direction == utilities::Direction::LONG) ? 1 : -1;
    auto *lc = option_data.at("long_call");
    auto *sc = option_data.at("short_call");
    auto *sp = option_data.at("short_put");
    auto *lp = option_data.at("long_put");
    std::vector<utilities::Leg> legs = {
        create_leg(*lc, sign > 0 ? utilities::Direction::LONG : utilities::Direction::SHORT, volume),
        create_leg(*sc, sign > 0 ? utilities::Direction::SHORT : utilities::Direction::LONG, volume),
        create_leg(*sp, sign > 0 ? utilities::Direction::SHORT : utilities::Direction::LONG, volume),
        create_leg(*lp, sign > 0 ? utilities::Direction::LONG : utilities::Direction::SHORT, volume),
    };
    return {legs, generate_combo_signature(legs)};
}

auto ComboBuilderEngine::generate_combo_signature(const std::vector<utilities::Leg>& legs) -> std::string {
    std::vector<std::string> parts;
    for (const auto& leg : legs) {
        std::string sym = leg.symbol.value_or("");
        if (sym.empty()) { continue;
}
        size_t pos = 0;
        std::vector<std::string> tokens;
        for (size_t i = 0; i <= sym.size(); ++i) {
            if (i == sym.size() || sym[i] == '-') {
                if (i > pos) { tokens.push_back(sym.substr(pos, i - pos));
}
                pos = i + 1;
            }
        }
        if (tokens.size() >= 4) {
            parts.push_back(tokens[1] + tokens[2] + tokens[3]);
}
    }
    std::ranges::sort(parts);
    std::string result;
    for (size_t i = 0; i < parts.size(); ++i) {
        if (i != 0U) { result += "-";
}
        result += parts[i];
    }
    return result;
}

}  // namespace engines
